### PR TITLE
Add update Drush command to fix RELS-INT inconsistencies.

### PR DESCRIPTION
**Pull request by jordandukart**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2067

# What does this Pull Request do?
Adds a Drush command that identifies potential inconsistencies in RELS-INT statements and XACML policies.

# What's new?
An update hook and a Drush command to identify and or fix RELS-INT statements that may be missing.

# How should this be tested?
Use the Tuque commit before https://github.com/Islandora/tuque/pull/161.
Create an XACML policy that specifies multiple datastream rules.
Run the Drush command as listed in the pull request and note that an inconsistency is identified and it fixes it.
Running the Drush command again does not find the object.

Example:
* Does this change require documentation to be updated? No.
* Does this change add any new dependencies?  No.
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)?  The Drush command needs to be ran so any potential broken POLICYs are fixed.
* Could this change impact execution of existing code? Makes broken POLICYs work as intended.

# Interested parties
@DiegoPino

### CODE REVIEW

**Comment by jordandukart** [comment]
Depends on https://github.com/Islandora/tuque/pull/161

File: api/islandora_xacml_api.drush.inc
Comment (adam-vessey) [inline api/islandora_xacml_api.drush.inc]: If `$to_update` is already `TRUE`, there's no real need to enter this loop... or even go to the next iteration of the containing loop. Would `break 2;` in both the internal loops make sense, to break out directly to [where `$to_update` is used](https://github.com/Islandora/islandora_xacml_editor/pull/138/files#diff-94b90cefd08aee17cbb778413a880ac0R283)?

File: api/islandora_xacml_api.drush.inc
Comment (adam-vessey) [inline api/islandora_xacml_api.drush.inc]: Should probably be `$sandbox['progress']`, as we haven't managed to hit `$sandbox['total']`...

Comment (adam-vessey) [reply api/islandora_xacml_api.drush.inc]: ... Not sure if we care to go fix [where this was copy/pasta'd from](https://github.com/Islandora/islandora_xacml_editor/pull/138/files#diff-94b90cefd08aee17cbb778413a880ac0R133)?

File: api/islandora_xacml_api.drush.inc
Comment (adam-vessey) [inline api/islandora_xacml_api.drush.inc]: Copy/pasta? We're not changing the content of the `POLICY` datastream...

File: api/islandora_xacml_api.drush.inc
Comment (adam-vessey) [inline api/islandora_xacml_api.drush.inc]: `s/dependenies/dependencies/`

Comment (adam-vessey) [reply api/islandora_xacml_api.drush.inc]: [Delicious copypasta](https://github.com/Islandora/islandora_xacml_editor/pull/138/files#diff-94b90cefd08aee17cbb778413a880ac0R22).

**Comment by DiegoPino** [comment]
Once tested and all concerns addressed, this pull (drush command) needs to go into release notes. @rosiel reminder in case I loose the document where I'm putting urgent notes for this release!

File: api/islandora_xacml_api.drush.inc
Comment (adam-vessey) [inline api/islandora_xacml_api.drush.inc]: Given the separate `$commit` bit... doesn't quite make sense to say "Updated" if nothing was actually done to the object...

**Review by adam-vessey — COMMENTED** [review]
One more little thing; otherwise, looks good.

File: api/islandora_xacml_api.drush.inc
Comment (adam-vessey) [inline api/islandora_xacml_api.drush.inc]: Should probably reflect `$limit`.

**Review by adam-vessey — APPROVED** [review]

**Comment by DiegoPino** [comment]
@adam-vessey @jordandukart will give this one a spin with real data and if all goes right, merging tomorrow to give folks some time to comment. Thanks!

**Comment by jordandukart** [comment]
Have you had a chance to look at this yet @DiegoPino ?

**Comment by DiegoPino** [comment]
Almost there. Seems to work ok but want to be sure i can test it with actual bad objects and a lot of them. Will be done during the day.

thanks

File: api/islandora_xacml_api.install
Comment (DiegoPino) [inline api/islandora_xacml_api.install]: This is not an issue but more like something we need to discuss. having islandora_xacml_api_save_relationships disabled at the moment of running this script does not mean that it was always so, which could still mean that wrong RELS-INT properties could be in place.  Not a big issue, but i see it also on the drush script too....
https://github.com/Islandora/islandora_xacml_editor/pull/138/files#diff-94b90cefd08aee17cbb778413a880ac0R186
I mean, ok, don't run the test automatically (maybe?) if that option is disabled , but if i suspect/want to run it, then line R186 will not allow me to run it at all?  We can discuss this on today's call of course.

File: api/islandora_xacml_api.drush.inc
Comment (DiegoPino) [inline api/islandora_xacml_api.drush.inc:223]: I know this is right, but having the return in the middle of $query variable assignment is complicated to read/follow. Do you think thole filter creation logic his could be moved out/before the actual assignment and then added to the query in a cleaner more readable way? Line 234 is especially confusing since the `$query` is formatted before (in our minds of course) $query is actually finished to be assigned. I know you need the different $query to be count. It's correct, just not easy to read.

File: api/islandora_xacml_api.drush.inc
Comment (DiegoPino) [inline api/islandora_xacml_api.drush.inc:197]: could  ?ds2 <http://islandora.ca/ontology/relsint#isViewableByRole> "administrator" UNION ?ds2 <http://islandora.ca/ontology/relsint#isViewableByUser> "fedoraAdmin" exclude objects?

File: api/islandora_xacml_api.drush.inc
Comment (DiegoPino) [inline api/islandora_xacml_api.drush.inc:229]: I don't completely understand the $filter logic here. Will try to dissect it and please correct me:
1. $query fetches all objects that have an attached policy and any indexed data stream with those two rels int properties (union) in case only one exists.

2.  at the end , limit is set to $LIMIT,the number of objects to process we pass to drush from drush_get_option('limit', 100),.  Lets say 100.
So it will get me every time 100.
3. The filter is set (in case it runs for the second time) to all objects created after $sandbox['last_date'] (accumulated on each batch run) OR   date is the same as $sandbox['last_date'] and the PID is larger/higher/next to the last PID processed? What makes you assume that PID will be actually higher if you are ordering by datecreate? and/or using UUID?
3. You run also an un-limited query (no $limit statement?) with the filter applied version at line 214 that gets us all (count) objects that match the query. 

So regarding the $filter and point 3(now i know what i don't get) not sure if checking the creationdate and the comparing a string via >pid is enough. Also, do all of our CMODELS get XACML inherited? 
Maybe dealing with modification dates.. in case the actual "fix" is being applied and it is not a dry run.

File: api/islandora_xacml_api.drush.inc
Comment (DiegoPino) [inline api/islandora_xacml_api.drush.inc:268]: only isViewableByRole and isViewableByUser was affected? any other predicate?

**Review by DiegoPino — COMMENTED** [review]
@jordandukart @adam-vessey added a few comments/questions and concerns. Not directly requesting changes right now so opening to your criteria also. Running this now on real data to see how it behaves. Could you give me some feedback on my questions? @jonathangreen pinging you here also since you have been quite active on committing to this module. Thanks!

**Comment by jordandukart** [comment]
Just saw these @DiegoPino will try and comment before our call.

File: api/islandora_xacml_api.drush.inc
Comment (jordandukart) [reply api/islandora_xacml_api.drush.inc:197]: Well 'fedoraAdmin' and 'administrator' are the user/roles that are defined to must be present when creating XACML policies. It'd be possible that maybe one is missing with our RELS-INT issues hence the UNION: https://github.com/Islandora/islandora_xacml_editor/blob/7.x/api/includes/xacml.inc#L277-L285.

File: api/islandora_xacml_api.drush.inc
Comment (jordandukart) [reply api/islandora_xacml_api.drush.inc:229]: Missing an additional order by `?pid` in the base query, should be `ORDER BY ?created ?pid` that is in the event that the ISO8601 stamp matches directly we'll fallback on the PID being ahead of the last one processed.

Not sure I fully understand your question(s) here but I'll lay out how this batch operates:

1. Batch is initiated and the total amount of objects that are to be processed is grabbed mainly for progress reporting / handling the end condition for the batch.
2. Initial chunk of objects are grabbed (without any filters) and processed, each storing their created date in the sandbox.
3. Initial chunk completes, next result set is grabbed using the created date of the last object processed for comparison purposes.

And so on.

File: api/islandora_xacml_api.drush.inc
Comment (jordandukart) [reply api/islandora_xacml_api.drush.inc:268]: These are just identifying conditions based on the defaults set in https://github.com/Islandora/islandora_xacml_editor/blob/7.x/api/includes/xacml.inc#L277-L285. RELS-INT has no isManageableBy predicate as per https://jira.duraspace.org/browse/ISLANDORA-1273.

File: api/islandora_xacml_api.install
Comment (jordandukart) [reply api/islandora_xacml_api.install]: Open for discussion sure.

File: api/islandora_xacml_api.drush.inc
Comment (jordandukart) [reply api/islandora_xacml_api.drush.inc:223]: I mean we could make a $count_query variable on 216 if you feel like that'll improve readability.

File: api/islandora_xacml_api.drush.inc
Comment (DiegoPino) [reply api/islandora_xacml_api.drush.inc:229]: Hi, cool!. All good, the other 1-4 (except 3) were just my own dissecting to understand the hows and also to let other people reading get why my concern in 3. order by pid should solve the issue.

File: api/islandora_xacml_api.drush.inc
Comment (DiegoPino) [reply api/islandora_xacml_api.drush.inc:223]: I feel if we could simply assign in one block (no limit not filter), then apply conditional logic to build the actual filter and/or put the limit  after. Then simple append strings at the end or as you said $count_query. This is just pickiness so if any other @Islandora/7-x-1-x-committers feels readability is fine, we can leave it like that.

File: api/islandora_xacml_api.install
Comment (DiegoPino) [reply api/islandora_xacml_api.install]: Ok, in the agenda for today. If this passes as is, we should be ready to merge, if not we could just remove the internal condition (inside the actual batch process) to allow people to run this independently of their current selection

File: api/islandora_xacml_api.drush.inc
Comment (DiegoPino) [reply api/islandora_xacml_api.drush.inc:268]: Ok, that answers my question. cool

**Comment by jordandukart** [comment]
Hey @DiegoPino what was the actionable outcome from the call yesterday?

**Comment by DiegoPino** [comment]
Not sure? What about just adding a notice so people know that in order to run the drush script they need to have the policy to rels-int toggle enabled (which is by default enabled I know)?. I did not read/ hear too much community reaction and I really want to merge this today.  Also, in any case, this will go into the release notes. so your call. If you feel that change can be done, happy, if not ping me and let's merge. thanks

**Comment by jordandukart** [comment]
K, I'll add a conditional notice in the update hook.

**Comment by jordandukart** [comment]
Updated the phrasing, thoughts @DiegoPino ?

**Comment by DiegoPino** [comment]
Reads fine. Thanks. Will wait for Travis to finish and then merge. Feel free to make the corresponding 7.x-1.10 release pull request. great work

**Pull request merged**
